File: docs/concepts/chapters/leblanc/lora/leblanc/poses/repaint.mjs
#!/usr/bin/env node
/**
 * Masked repaint of one Leblanc pose render (FFX-2 only, Chapter 6): the follow-up the
 * round-3 judge named ("masked repaints, the tool that already worked here",
 * sets/ormi/round3/repaint-heart.mjs), for a defect the words and the pose control did
 * not fix. Everything outside the mask is composited back pixel for pixel
 * (ImageCompositeMasked); no IP-Adapter, no ControlNet.
 *
 *   node docs/concepts/chapters/leblanc/lora/leblanc/poses/repaint.mjs <state> <tag> <name>
 *        --ellipse cx,cy,rx,ry[,angle] [--ellipse ...] --seeds a,b --denoise 1 --lora 0.5
 *        --positive "..." --negative "..." [--inpaint 1]
 *
 * <tag> is D:/Tools/pyrefly-lora/leblanc/poses/<tag>.raw.png; ellipses are in raw
 * pixels. --inpaint 1 uses VAEEncodeForInpaint (fills the hole from scratch, denoise 1);
 * otherwise VAEEncode + SetLatentNoiseMask at --denoise. Writes <tag>.<name><seed>.raw.png,
 * the cut-out, the mask and a sidecar beside the source.
 */
import { writeFileSync, readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { spawnSync } from 'node:child_process';

const HERE = dirname(fileURLToPath(import.meta.url));
const REPO = resolve(HERE, '..', '..', '..', '..', '..', '..', '..');
const G = await import(pathToFileURL(join(REPO, 'tools/gen/comfy.mjs')).href);
const C = await import(pathToFileURL(join(REPO, 'tools/gen/cutout-guard.mjs')).href);
const R = await import(pathToFileURL(join(HERE, 'render.mjs')).href);
const { stageImage, cutout, STYLE_TAGS, QUALITY_TAGS, BASE_NEGATIVE } = G;
const CKPT = 'animagine-xl-4.0-opt.safetensors';

const argv = process.argv.slice(2);
const [state, tag, name] = argv;
const opt = { ellipse: [] };
for (let i = 3; i < argv.length; i += 2) {
  const k = argv[i].slice(2);
  if (k === 'ellipse') opt.ellipse.push(argv[i + 1].split(',').map(Number));
  else opt[k] = argv[i + 1];
}
const dir = R.OUT;
const src = join(dir, `${tag}.raw.png`);
const seeds = String(opt.seeds || '970001').split(',').map(Number);
const denoise = Number(opt.denoise ?? 1);
const lora = Number(opt.lora ?? 0.5);
const inpaint = opt.inpaint === '1';
const positive = `leblancX2, ${opt.positive}, ${STYLE_TAGS}, ${QUALITY_TAGS}`;
const negative = `${BASE_NEGATIVE}, ${opt.negative || ''}`;
const mask = join(dir, `${tag}.${name}.mask.png`);

const py = spawnSync('python', ['-c', `
from PIL import Image, ImageDraw, ImageFilter
import json, sys
im = Image.open(sys.argv[1]); S = 4
m = Image.new("L", im.size, 0)
for e in json.loads(sys.argv[3]):
    cx, cy, rx, ry = e[:4]; ang = e[4] if len(e) > 4 else 0
    layer = Image.new("L", (im.width * S, im.height * S), 0)
    ImageDraw.Draw(layer).ellipse([(cx-rx)*S, (cy-ry)*S, (cx+rx)*S, (cy+ry)*S], fill=255)
    layer = layer.rotate(ang, center=(cx*S, cy*S)).resize(im.size, Image.LANCZOS)
    m = Image.fromarray(__import__("numpy").maximum(__import__("numpy").array(m), __import__("numpy").array(layer)))
m = m.filter(ImageFilter.GaussianBlur(3))
Image.merge("RGB", (m, m, m)).save(sys.argv[2])
`, src, mask, JSON.stringify(opt.ellipse)], { encoding: 'utf8' });
if (py.status !== 0) throw new Error(py.stderr);

const img = stageImage(src);
const msk = stageImage(mask);
for (const seed of seeds) {
  const g = {
    4: { class_type: 'CheckpointLoaderSimple', inputs: { ckpt_name: CKPT } },
    10: { class_type: 'LoraLoader', inputs: { model: ['4', 0], clip: ['4', 1], lora_name: 'leblanc-x2.safetensors', strength_model: lora, strength_clip: lora } },
    6: { class_type: 'CLIPTextEncode', inputs: { text: positive, clip: ['10', 1] } },
    7: { class_type: 'CLIPTextEncode', inputs: { text: negative, clip: ['10', 1] } },
    11: { class_type: 'LoadImage', inputs: { image: img, upload: 'image' } },
    12: { class_type: 'LoadImage', inputs: { image: msk, upload: 'image' } },
    13: { class_type: 'ImageToMask', inputs: { image: ['12', 0], channel: 'red' } },
    3: {
      class_type: 'KSampler',
      inputs: { seed, steps: 28, cfg: 6, sampler_name: 'euler_ancestral', scheduler: 'normal', denoise: inpaint ? 1 : denoise, model: ['10', 0], positive: ['6', 0], negative: ['7', 0], latent_image: inpaint ? ['14', 0] : ['15', 0] },
    },
    8: { class_type: 'VAEDecode', inputs: { samples: ['3', 0], vae: ['4', 2] } },
    16: { class_type: 'ImageCompositeMasked', inputs: { destination: ['11', 0], source: ['8', 0], x: 0, y: 0, resize_source: false, mask: ['13', 0] } },
    9: { class_type: 'SaveImage', inputs: { filename_prefix: `pyrefly/lora-leblanc-repaint-${state}`, images: ['16', 0] } },
  };
  if (inpaint) g[14] = { class_type: 'VAEEncodeForInpaint', inputs: { pixels: ['11', 0], vae: ['4', 2], mask: ['13', 0], grow_mask_by: 8 } };
  else {
    g[17] = { class_type: 'VAEEncode', inputs: { pixels: ['11', 0], vae: ['4', 2] } };
    g[15] = { class_type: 'SetLatentNoiseMask', inputs: { samples: ['17', 0], mask: ['13', 0] } };
  }
  const out = `${tag}.${name}${seed}`;
  const secs = await R.run(g, join(dir, `${out}.raw.png`));
  const s = R.STATES[state];
  const cut = cutout(join(dir, `${out}.raw.png`), join(dir, `${out}.png`));
  const guard = await C.checkCutoutFile(join(dir, `${out}.png`), { sourceWidth: s.size[0], sourceHeight: s.size[1], composition: s.composition });
  const base = JSON.parse(readFileSync(join(dir, `${tag}.json`), 'utf8'));
  writeFileSync(join(dir, `${out}.json`), JSON.stringify({
    ...base, tag: out, cutout: cut, guard: { ok: guard.ok, reasons: guard.reasons },
    repaint: { from: `${tag}.raw.png`, name, seed, ellipses: opt.ellipse, feather: 3, inpaint, denoise: inpaint ? 1 : denoise, lora, positive, negative, seconds: secs },
  }, null, 1));
  console.log(`[leblanc-repaint] ${out} ${secs.toFixed(1)} s guard ${guard.ok ? 'ok' : 'REJECT ' + guard.reasons.join('; ')}`);
}

File: docs/concepts/chapters/macalania/r3-poses/scripts/repaint.py
"""Masked repaint of the HOLES only (FFX only, METHOD-CHECK step 2.5): crop the hole's box plus a margin,
RealESRGAN x4 then Lanczos to a 1024 short side, Animagine XL 4.0 Opt img2img with SetLatentNoiseMask over the
hole (dilated) only, N seeds in ONE prompt; back down, composite inside the mask only (feathered), then the
protected pixels (the moved idle part) are pasted back exactly. Shared-GPU rule: /queue empty 3 minutes first.
usage: repaint.py <draft.png> <hole.png> <protect.png|-> <out_prefix> <denoise> <seeds,comma> "<pos>" "<neg>" [margin] [dilate]"""
import sys, json, time, uuid, urllib.request, urllib.parse, os
import numpy as np, cv2
COMFY = 'http://127.0.0.1:8188'
src, holep, protp, outp, den, seeds, pos, neg = sys.argv[1:9]
margin = int(sys.argv[9]) if len(sys.argv) > 9 else 48
dil = int(sys.argv[10]) if len(sys.argv) > 10 else 5
den = float(den); seeds = [int(s) for s in seeds.split(',')]
im = cv2.imread(src, -1).astype(np.float32) / 255
hole = cv2.imread(holep, 0) > 0
prot = (cv2.imread(protp, 0) > 0) if protp != '-' else np.zeros_like(hole)
m = cv2.dilate(hole.astype(np.uint8), np.ones((2 * dil + 1, 2 * dil + 1), np.uint8)) > 0
m &= (im[..., 3] > 0.5) & ~prot
ys, xs = np.nonzero(m)
x0, y0 = max(0, xs.min() - margin), max(0, ys.min() - margin)
x1, y1 = min(im.shape[1], xs.max() + margin + 1), min(im.shape[0], ys.max() + margin + 1)
crop = im[y0:y1, x0:x1]; mc = m[y0:y1, x0:x1].astype(np.float32)
a = crop[..., 3:4]; rgb = crop[..., :3] * a + 0.42 * (1 - a)
s = 1024 / min(crop.shape[:2]); W = int(round(crop.shape[1] * s / 8) * 8); H = int(round(crop.shape[0] * s / 8) * 8)
mbig = cv2.GaussianBlur(cv2.resize(mc, (W, H), interpolation=cv2.INTER_LINEAR), (0, 0), 3)


def get(path):
    return json.loads(urllib.request.urlopen(COMFY + path, timeout=60).read())


def upload(arr, name):
    ok, buf = cv2.imencode('.png', arr); b = uuid.uuid4().hex
    body = (f'--{b}\r\nContent-Disposition: form-data; name="image"; filename="{name}"\r\nContent-Type: image/png\r\n\r\n').encode() \
        + buf.tobytes() + f'\r\n--{b}\r\nContent-Disposition: form-data; name="overwrite"\r\n\r\ntrue\r\n--{b}--\r\n'.encode()
    r = urllib.request.Request(COMFY + '/upload/image', data=body, headers={'Content-Type': f'multipart/form-data; boundary={b}'})
    return json.loads(urllib.request.urlopen(r, timeout=60).read())['name']


tag = os.path.basename(outp)
iname = upload(np.clip(rgb * 255, 0, 255).astype(np.uint8), f'mac-r3p-{tag}-init.png')
mname = upload(np.clip(np.dstack([mbig] * 3) * 255, 0, 255).astype(np.uint8), f'mac-r3p-{tag}-mask.png')
wf = {
    '1': {'class_type': 'CheckpointLoaderSimple', 'inputs': {'ckpt_name': 'animagine-xl-4.0-opt.safetensors'}},
    '2': {'class_type': 'LoadImage', 'inputs': {'image': iname}},
    '3': {'class_type': 'UpscaleModelLoader', 'inputs': {'model_name': 'RealESRGAN_x4plus.pth'}},
    '4': {'class_type': 'ImageUpscaleWithModel', 'inputs': {'upscale_model': ['3', 0], 'image': ['2', 0]}},
    '5': {'class_type': 'ImageScale', 'inputs': {'image': ['4', 0], 'upscale_method': 'lanczos', 'width': W, 'height': H, 'crop': 'disabled'}},
    '6': {'class_type': 'VAEEncode', 'inputs': {'pixels': ['5', 0], 'vae': ['1', 2]}},
    '7': {'class_type': 'LoadImage', 'inputs': {'image': mname}},
    '8': {'class_type': 'ImageToMask', 'inputs': {'image': ['7', 0], 'channel': 'red'}},
    '9': {'class_type': 'SetLatentNoiseMask', 'inputs': {'samples': ['6', 0], 'mask': ['8', 0]}},
    '10': {'class_type': 'CLIPTextEncode', 'inputs': {'text': pos, 'clip': ['1', 1]}},
    '11': {'class_type': 'CLIPTextEncode', 'inputs': {'text': neg, 'clip': ['1', 1]}},
}
for i, sd in enumerate(seeds):
    k, d, sv = str(100 + 3 * i), str(101 + 3 * i), str(102 + 3 * i)
    wf[k] = {'class_type': 'KSampler', 'inputs': {'seed': sd, 'steps': 28, 'cfg': 6, 'sampler_name': 'euler_ancestral', 'scheduler': 'normal',
                                                  'denoise': den, 'model': ['1', 0], 'positive': ['10', 0], 'negative': ['11', 0], 'latent_image': ['9', 0]}}
    wf[d] = {'class_type': 'VAEDecode', 'inputs': {'samples': [k, 0], 'vae': ['1', 2]}}
    wf[sv] = {'class_type': 'SaveImage', 'inputs': {'images': [d, 0], 'filename_prefix': f'pyrefly/mac-r3p-{tag}-s{sd}'}}
calm = None
while True:
    q = get('/queue')
    if q['queue_running'] or q['queue_pending']:
        calm = None
    elif calm is None:
        calm = time.time()
    if calm and time.time() - calm >= 180:
        break
    time.sleep(10)
r = urllib.request.Request(COMFY + '/prompt', data=json.dumps({'prompt': wf, 'client_id': 'mac-r3-poses'}).encode(), headers={'Content-Type': 'application/json'})
pid = json.loads(urllib.request.urlopen(r, timeout=60).read())['prompt_id']
while True:
    h = get('/history/' + pid)
    if pid in h and h[pid].get('status', {}).get('completed'):
        break
    if pid in h and h[pid].get('status', {}).get('status_str') == 'error':
        raise SystemExit('comfy error ' + json.dumps(h[pid]['status'])[:1500])
    time.sleep(3)
e = h[pid]; msgs = e.get('status', {}).get('messages', [])
t = {mm[0]: mm[1].get('timestamp') for mm in msgs}
secs = ((t.get('execution_success') or 0) - (t.get('execution_start') or 0)) / 1000
kk = cv2.GaussianBlur(mc, (0, 0), 1.2)[..., None]
res_meta = []
for node, o in e['outputs'].items():
    for img in o.get('images', []):
        sd = int(img['filename'].split('-s')[-1].split('_')[0])
        q = urllib.parse.urlencode({'filename': img['filename'], 'subfolder': img.get('subfolder', ''), 'type': img.get('type', 'output')})
        gen = cv2.imdecode(np.frombuffer(urllib.request.urlopen(COMFY + '/view?' + q).read(), np.uint8), cv2.IMREAD_COLOR).astype(np.float32) / 255
        black = float(gen.mean()) < 0.02
        small = cv2.resize(gen, (x1 - x0, y1 - y0), interpolation=cv2.INTER_AREA)
        res = im.copy()
        res[y0:y1, x0:x1, :3] = crop[..., :3] * (1 - kk) + small * kk
        res[m, 3] = 1.0
        res[prot] = im[prot]
        cv2.imwrite(f'{outp}.s{sd}.png', np.clip(res * 255 + 0.5, 0, 255).astype(np.uint8))
        res_meta.append(dict(seed=sd, black=black, mean=float(gen.mean())))
meta = dict(prompt_id=pid, seconds=secs, denoise=den, box=[int(x0), int(y0), int(x1), int(y1)], canvas=[W, H], maskPx=int(m.sum()), pos=pos, neg=neg, outputs=res_meta)
json.dump(meta, open(outp + '.repaint.json', 'w'), indent=1)
with open(os.path.join(os.path.dirname(os.path.abspath(outp)), 'gpu-log.tsv'), 'a') as f:
    f.write(f'{time.strftime("%Y-%m-%dT%H:%M:%S")}\t{pid}\t{tag}\t{secs:.1f}\n')
print(json.dumps(dict(prompt_id=pid, seconds=secs, box=meta['box'], outputs=res_meta)))
if any(rm['black'] for rm in res_meta):
    print('ALL-BLACK FRAME: bad GPU state, stop'); sys.exit(3)
